jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c

```diff
@@ -1890,19 +1890,49 @@ ecma_builtin_array_prototype_object_last_index_of (ecma_value_t this_arg, /**< t
   return ret_value;
 } /* ecma_builtin_array_prototype_object_last_index_of */
 
+
 /**
- * The Array.prototype object's 'every' routine
- *
- * See also:
- *          ECMA-262 v5, 15.4.4.16
- *
- * @return ecma value
- *         Returned value must be freed with ecma_free_value.
+ * If the argument is convertible to boolean true, returns ECMA_VALUE_TRUE,
+ * otherwise returns ECMA_VALUE_EMPTY.
  */
 static ecma_value_t
-ecma_builtin_array_prototype_object_every (ecma_value_t this_arg, /**< this argument */
-                                           ecma_value_t arg1, /**< callbackfn */
-                                           ecma_value_t arg2) /**< thisArg */
+ecma_builtin_convert_true_maybe (ecma_value_t var)
+{
+  if (ecma_op_to_boolean (var))
+  {
+    return ECMA_VALUE_TRUE;
+  }
+
+  return ECMA_VALUE_EMPTY;
+} /* ecma_builtin_convert_true_maybe */
+
+/**
+ * If the argument is convertible to boolean false, returns ECMA_VALUE_FALSE,
+ * otherwise returns ECMA_VALUE_EMPTY.
+ */
+static ecma_value_t
+ecma_builtin_convert_false_maybe (ecma_value_t var)
+{
+  if (!ecma_op_to_boolean (var))
+  {
+    return ECMA_VALUE_FALSE;
+  }
+
+  return ECMA_VALUE_EMPTY;
+} /* ecma_builtin_convert_false_maybe */
+
+/**
+ * Applies the provided function to each element of the array as long as
+ * the return value stays undefined. If the return value stays undefined
+ * after processing the whole array, a default value can be returned.
+ */
+static ecma_value_t
+ecma_builtin_array_apply (ecma_value_t this_arg, /**< this argument */
+                          ecma_value_t arg1, /**< callbackfn */
+                          ecma_value_t arg2, /**< thisArg */
+                          ecma_value_t (*check_stop_cb)(ecma_value_t), /**< retval update fn */
+                          ecma_value_t default_retval /**< default if empty */ )
```

Comment:
IMHO, I would avoid using the callback method to do the work. What we could do is to only pass an enum value (something like: ARRAY_EVERY, _SOME, _FOR_EACH) and based on this value we can calculate the stop and default return values.